src/JestExt.ts

```diff
@@ -59,121 +61,90 @@ export class JestExt {
-    // we can handle here differently.
-    if (this.jestProcess) {
-      this.jestProcess.closeProcess()
-      delete this.jestProcess
+    if (this.jestProcessManager.numberOfProcesses > 0) {
+      return
     }
-
-    let maxRestart = 4
-    this.jestProcess = new Runner(this.workspace)
-
-    this.jestProcess
-      .on('debuggerProcessExit', () => {
-        this.channel.appendLine('Closed Jest')
-
-        if (this.forcedClose) {
-          this.forcedClose = false
-          return
-        }
-
-        if (maxRestart-- <= 0) {
-          console.warn('jest has been restarted too many times, please check your system')
-          status.stopped('(too many restarts)')
-          return
-        }
-
-        this.closeJest()
-        this.startWatchMode()
-      })
-      .on('executableJSON', (data: JestTotalResults) => {
-        this.updateWithData(data)
-      })
-      .on('executableOutput', (output: string) => {
-        if (!this.shouldIgnoreOutput(output)) {
-          this.channel.appendLine(output)
-        }
-      })
-      .on('executableStdErr', (error: Buffer) => {
-        const message = error.toString()
-
-        if (this.shouldIgnoreOutput(message)) {
-          return
-        }
-
-        // The "tests are done" message comes through stdErr
-        // We want to use this as a marker that the console should
-        // be cleared, as the next input will be from a new test run.
-        if (this.clearOnNextInput) {
-          this.clearOnNextInput = false
-          this.parsingTestFile = false
-          this.testsHaveStartedRunning()
-        }
-        // thanks Qix, http://stackoverflow.com/questions/25245716/remove-all-ansi-colors-styles-from-strings
-        const noANSI = message.replace(
-          /[\u001b\u009b][[()#;?]*(?:[0-9]{1,4}(?:;[0-9]{0,4})*)?[0-9A-ORZcf-nqry=><]/g,
-          ''
-        )
-        if (noANSI.includes('snapshot test failed')) {
-          this.detectedSnapshotErrors()
-        }
-
-        this.channel.appendLine(noANSI)
-      })
-      .on('nonTerminalError', (error: string) => {
-        this.channel.appendLine(`Received an error from Jest Runner: ${error.toString()}`)
-      })
-      .on('exception', result => {
-        this.channel.appendLine(`\nException raised: [${result.type}]: ${result.message}\n`)
-      })
-      .on('terminalError', (error: string) => {
-        this.channel.appendLine('\nException raised: ' + error)
-      })
-
     // The theme stuff
     this.setupDecorators()
     // The bottom bar thing
     this.setupStatusBar()
     //reset the jest diagnostics
     resetDiagnostics(this.failDiagnostics)
 
-    this.forcedClose = false
-    // Go!
-    if (this.pluginSettings.runAllTestsFirst) {
-      this.jestProcess.start(false)
-    } else {
-      this.startWatchMode()
-    }
+    this.jestProcessManager.startJestProcess({
+      watch: true,
+      keepAlive: true,
+      exitCallback: (_, jestProcessInWatchMode) => {

```

Comment:
maybe I am missing something, this looks to me that all the jest output are only processed upon process exit... but during watch mode, the process will not exit isn't it? does the output get processed upon change?

Reply:
This is good one. Actually it is other way around. I have assigned the handlers correctly for the watch mode (the exit handler for the first process - the one running all the tests first - provides two arguments: one for the process that just exited, second for the jest process for the watch mode that just started), but I did not assigned them for the first process that runs all the tests first.

I have just fixed it - please see the most recent commit.